feat: support don't close on outside click of modal

Lalaluka opened the pull request:
Hey,
this is my suggestion for something mentioned in #428 
I'm unsure about the attribute Name. `closeClickOutside` doesn't sound that nice.

But basically, it adds an optional attribute, which enables Clicks outside the Modal to not close the modal.
Since someone might want to hook into this functionality I also added a new event `scaleCloseAttempt`. 
In the example of a form, some browser dialog could open and ask if you really want to close this modal or run some custom code saving the Modals Data in some lifted state before closing it. Of course, the event could just be ignored so the Modal just won't close.

I added an example in the index.html for people to test it.

The question is if something like this should also be added for the close button.

(this is just a suggestion. Feel free to just reject it if it's not an approach you like)

acstll commented:
Hey @Lalaluka 

this looks amazing (except for the attribute name `closeClickOutside` 😉).

> The question is if something like this should also be added for the close button.

I think so, this would make it more flexible, I had the idea you could `preventDefault` (or something similar) this previous event to avoid closing the modal.

I will try and properly review this tomorrow.

Thanks again!

Lalaluka commented:
I agree that the name is horrible I'll change it.
Also will try to adapt something similar for close Buttons and add it to the PR.

> I think so, this would make it more flexible, I had the idea you could preventDefault (or something similar) this previous event to avoid closing the modal.

Trying to understand your comment: So you want to be able to catch some kind of `beforeScaleClose` Event with the possibility of `preventdefault` to stop the propagation to the actual close Event. Looking at the code that might need a refactor of the eventing in my opinion. But if this is what you are suggesting I can try to make that happen.

acstll commented:
> I agree that the name is horrible I'll change it.
naming is hard :)

I would suggest using `backdropStatic`, based on #428 and default it to false. What do you think?

Regarding the `beforeScaleClose` idea, maybe it makes sense to do that separately, in another PR, because I agree it might need some refactoring as you mention. I'll do some quick research to clarify.

acstll commented:
By the way, I think the `scaleCloseAttempt` event name is fine 👍

acstll commented:
This is the pattern I was thinking about:
https://github.com/LeaVerou/awesomplete/blob/gh-pages/awesomplete.js#L285-L298

`dispatchEvent()` returns `false` if the event has been `preventDefault`ed — let's hope the Stencil emitter also returns… ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent#return_value))

So one could avoid closing the modal with something like this:

```js
modal.addEventListener('beforeScaleClose', (event) => { event.preventDefault() })
```

I remember seeing this somewhere, but I can't remember where.

Lalaluka commented:
Hi,
a bit of a late response but here is my solution.

Thanks for the Ideas @acstll ✌️

Reworking the eventing was easier than I thought. So I added a new Event:
`scaleBeforeClose` gets fired before close Events triggered by:

- Backdrop
- Pressing Escape
- Clicking the Close Icon

On any of those Actions, the Event gets fired. Calling `event.preventDefault()` in the Eventhandler enables the User to prevent the closure of the Modal. Alternativly some code can be executed before the modal closes. 
The Event Object carries a type in its `Details` field, which enables the identification of the trigger Action.
The types are:
`buttonClose` `escapePressed` and `backdrop`

The `scaleCloseAttempt` and the horribly named prop are removed. Since this Event can handle the same use-case.

I'm converting this from draft to Review since I'm quite happy with the outcome.
An example can be found in the index.html.

acstll commented:
Amazing! I'm taking a look later today 👀

acstll reviewed (changes requested):
Thank you @Lalaluka, I really like the implementation, it's really simple (and more flexible than the initial `backdrop-static` prop idea).

I'm suggesting making the types a bit more picky and renaming a method. What do you think?

You can also go ahead with the docs :)

acstll commented on packages/components/src/components/modal/modal.tsx:
```diff
@@ -109,6 +111,12 @@ export class Modal {
     }
   }
 
+  closeEvent(type: string) {
+    if (!this.scaleBeforeClose.emit(type).defaultPrevented) {
+      this.opened = false;
+    }
+  }
```
I was thinking I might have made up the "convention" thing because I didn't do any research within the project to confirm it (that we use objects in `event.detail`, so it might be more a personal thing 😁

Lalaluka replied:
Putting Objects into the details sounds rational since you can add other stuff later without breaking anything.

The rename is also good.

acstll replied:
I hadn't thought about the non-breaking outcome, now I like it even more. Anything "non-breaking" is good for us at this point :)

Lalaluka commented on packages/components/src/components/modal/modal.tsx:
```diff
@@ -69,6 +69,8 @@ export class Modal {
   @Event() scaleOpen?: EventEmitter;
   @Event() scaleClose?: EventEmitter;
+  /** Event firing before an Modal Action is called. Propagation to closing the Event can be stoped. Possible actions: `buttonClose` `escapePressed` `backdrop` */
+  @Event() scaleBeforeClose?: EventEmitter;
```
Was wondering if you have some kind of place to put types. Good addition.

acstll replied:
Usually at the top of the files, before the component code. I remember there used to be a "types" file for shared types, but I don't see it anymore…

Lalaluka commented:
I made the changes you requested and added documentation to the Events the Modal has.
Prior there was no documentation for the modal Events so I added all three.

acstll reviewed (approved):
Thank you!
